src/groupeddataframe/grouping.jl

```diff
@@ -177,6 +213,318 @@ end
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+"""
+    groupindices(gd::GroupedDataFrame)
+
+Return a vector of group indices for each row of `parent(gd)`.
+
+Rows appearing in group `gd[i]` are attributed index `i`. Rows not present in
+any group are attributed `missing` (this can happen if `skipmissing=true` was
+passed when creating `gd`, or if `gd` is a subset from a larger [`GroupedDataFrame`](@ref)).
+"""
+groupindices(gd::GroupedDataFrame) = replace(gd.groups, 0=>missing)
+
+"""
+    groupvars(gd::GroupedDataFrame)
+
+Return a vector of column names in `parent(gd)` used for grouping.
+"""
+groupvars(gd::GroupedDataFrame) = _names(gd)[gd.cols]
+
+# Get grouping variable index by its name
+function _groupvar_idx(gd::GroupedDataFrame, name::Symbol, strict::Bool)
+    i = findfirst(==(name), groupvars(gd))
+    i === nothing && strict && throw(ArgumentError("$name is not a grouping column"))
+    return i
+end
+
+"""
+    groupvalues([T::Type], gd::GroupedDataFrame)
+    groupvalues([T::Type], gd::GroupedDataFrame, i)
+    groupvalues(gd::GroupedDataFrame, i, col)
+
+Get the values of the grouping columns for `gd`, optionally specifying the group
+index `i` and grouping column `col`. `col` may be an integer index (of the
+grouping columns passed to [`groupby`](@ref)) or the column name itself. The
+optional first argument to the first two forms specifies the return type, which
+may be `NamedTuple` (default) or `Tuple`.
+
+### Returns
+
+Iterator over `Tuple`s/`NamedTuple`s, a single `Tuple`/`NamedTuple`, or a
+grouping column value.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(groupvalues(gd))
+6-element Array{NamedTuple{(:a, :b),Tuple{Symbol,Int64}},1}:
+ (a = :foo, b = 2)
+ (a = :bar, b = 1)
+ (a = :baz, b = 2)
+ (a = :foo, b = 1)
+ (a = :bar, b = 2)
+ (a = :baz, b = 1)
+
+julia> groupvalues(gd, 2)
+(a = :bar, b = 1)
+
+julia> groupvalues(gd, 2, 1)
+:bar
+
+julia> groupvalues(gd, 2, :a)
+:bar
+
+julia> groupvalues(Tuple, gd, 2)
+(:bar, 1)
+```
+"""
+groupvalues(T::Type, gd::GroupedDataFrame) = (groupvalues(T, gd, i) for i in 1:length(gd))
```

Comment:
The consequence of being lazy here (generator) is that it will be slow as `groupvalues(T, gd, i)` has to perform repeatedly the same work for each `i`. Maybe let us return a `Vector` to make this faster? (then in the docs we will specify that we return a vector)? Do you have an opinion here?

Reply:
Not disagreeing that an array would be better here, but why would the performance be worse with the generator?

Reply:
because `groupvalues(T, gd, i)` needs to determine the type of the element each time it is called. See e.g.:
```
NamedTuple{Tuple(groupvars(gd))}(groupvalues(Tuple, gd, i))
```
and you have to call `groupvars(gd)` in each iteration which adds some cost.

Also a generator does not work well with type inference (the caller will not know `eltype` even after using function barier).

Of course if we switched to array we have a tricky case that rows can have heterogeneous types so this means that the type of `Tuple` and `NamedTuple` will not be constant. And another level of complexity is that then we hit a distinction that `Tuple` is covariant and `NamedTuple` is invariant (but I guess we should default to standard Julia promotion rules here).

So - in general the case is not that easy to do it right. Therefore I do not push for this change very strongly. As generator is essentially less functional than a vector we can make this change later if we want to (but maybe you and @nalimilan come to the conclusion that it is better to do it in this PR).

Reply:
You mean a custom lazy `AbstractVector`? There are other advantages to that see [above thread](https://github.com/JuliaData/DataFrames.jl/pull/1908/files#r337931263). Regarding the computation of the element type, I guess we should just use the eltype of each column vector.

Reply:
I would just use `Vector`.

> I guess we should just use the eltype of each column vector.

You mean for the `eltype` of the vector? - this should be OK for `Tuple` as it is covariant, but it will be problematic for `NamedTuple` because it is invariant, so the pushed `NamedTuples` would get `convert`ed to a supertype.

@jlumpe - sorry for this discussion (without clear conclusions and considering various options), but at least for me, this part of the design is tricky to get right.

Reply:
Well a `Vector` of names tuples is generally not what people need. Also that requires making a copy, which is probably not very useful. We could almost return a `StructArray`.

> You mean for the `eltype` of the vector? - this should be OK for `Tuple` as it is covariant, but it will be problematic for `NamedTuple` because it is invariant, so the pushed `NamedTuples` would get `convert`ed to a supertype.

Pushed to what? If the vector contains named tuples of the same type (the eltype of the columns), I don't see when the problem would happen.

Reply:
If the `eltype` of column is `Integer` for example here you have what happens:
```
julia> v = NamedTuple{(:a,),Tuple{Integer}}[]
0-element Array{NamedTuple{(:a,),Tuple{Integer}},1}

julia> z = (a=1,)
(a = 1,)

julia> push!(v, z)
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 NamedTuple{(:a,),Tuple{Integer}}((1,))

julia> v[1]
NamedTuple{(:a,),Tuple{Integer}}((1,))

julia> typeof(z)
NamedTuple{(:a,),Tuple{Int64}}
```
so you get a silent promotion.

Reply:
I have written about `push!` but essentially it is an issue of storage:
```
julia> v = Vector{NamedTuple{(:a,),Tuple{Integer}}}(undef, 1)
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 #undef

julia> v[1] = z
(a = 1,)

julia> v
1-element Array{NamedTuple{(:a,),Tuple{Integer}},1}:
 NamedTuple{(:a,),Tuple{Integer}}((1,))
```
is the same

Reply:
in summary - maybe for the time being it is OK to stay with the generator and resolve if we want something more fancy in a separate PR. I am not sure what is best.

Reply:
But why would promotion be a problem? Anyway that vector would be read-only.

The reason why I suggest using a vector is that it would allow us to avoid adding `groupvalues`, and use `keys` directly instead.

Reply:
So can you sumarize what you propose please? Is it to make `keys` return `Vector{GroupKey}` and then the user can just call `Tuple` or `NamedTuple` on it if one wants to get a specific type (similarly to what we currently have with `DataFrameRow`)? It should have the same performance as what `groupvalues` now has. Is this your idea. If yes I think it is OK.

Reply:
I was thinking of having `keys` return a custom `AbstractVector` which would lazily create `GroupKey` objects on indexing.

Reply:
This is also OK (I understand that this custom `AbstractVector` would be also read-only). This should not be that hard (we do a similar thing e.g. for `eachrow`), we just have to remember to follow https://docs.julialang.org/en/latest/manual/interfaces/#man-interface-array-1 (which seems easy, but then usually some corner cases pop up 😄).